warp/src/raygun/community.rs

```diff
@@ -0,0 +1,224 @@
+use chrono::{DateTime, Utc};
+use indexmap::{IndexMap, IndexSet};
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use crate::crypto::DID;
+use crate::raygun::Error;
+
+pub type Role = String;
+pub type CommunityPermissions = IndexMap<CommunityPermission, IndexSet<Role>>;
+pub type CommunityChannelPermissions = IndexMap<CommunityChannelPermission, IndexSet<Role>>;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Community {
+    id: Uuid,
+    name: String,
+    description: Option<String>,
+    creator: DID,
+    created: DateTime<Utc>,
+    modified: DateTime<Utc>,
+    members: IndexSet<DID>,
+    channels: IndexSet<Uuid>,
+    roles: IndexSet<Role>,
+    permissions: CommunityPermissions,
+}
+impl Community {
+    pub fn id(&self) -> Uuid {
+        self.id
+    }
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+    pub fn description(&self) -> &Option<String> {
+        &self.description
+    }
+    pub fn creator(&self) -> &DID {
+        &self.creator
+    }
+    pub fn created(&self) -> DateTime<Utc> {
+        self.created
+    }
+    pub fn modified(&self) -> DateTime<Utc> {
+        self.modified
+    }
+    pub fn members(&self) -> &IndexSet<DID> {
+        &self.members
+    }
+    pub fn channels(&self) -> &IndexSet<Uuid> {
+        &self.channels
+    }
+    pub fn roles(&self) -> &IndexSet<Role> {
+        &self.roles
+    }
+    pub fn permissions(&self) -> &CommunityPermissions {
+        &self.permissions
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CommunityChannel {
+    id: Uuid,
+    name: String,
+    description: Option<String>,
+    created: DateTime<Utc>,
+    modified: DateTime<Utc>,
+    channel_type: CommunityChannelType,
+    permissions: CommunityChannelPermissions,
+}
+
+impl CommunityChannel {
+    pub fn id(&self) -> Uuid {
+        self.id
+    }
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+    pub fn description(&self) -> &Option<String> {
+        &self.description
+    }
+    pub fn created(&self) -> DateTime<Utc> {
+        self.created
+    }
+    pub fn modified(&self) -> DateTime<Utc> {
+        self.modified
+    }
+    pub fn channel_type(&self) -> CommunityChannelType {
+        self.channel_type
+    }
+    pub fn permissions(&self) -> &CommunityChannelPermissions {
+        &self.permissions
+    }
+}
+#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
+pub enum CommunityChannelType {
+    Text,
+    Voice,
+}
```

Comment:
Not sure if we should have "Text" and "Voice" as a type since `RayGun` would be mostly for chat. We could probably let the frontend impl handle that in some way or maybe change `Voice` to something else that represents a non-text chat.

Reply:
From warp's side, Text and Voice channels are really both just Text channels and behave exactly the same. (Voice channels still need a chat in them) We just need a way to store the 'intention' the channel was created for in Warp. That way the ui can know that a particular channel is intended to be voice enabled and also to know to render the chat differently. (Like a Discord Voice channel's embedded text chat)

Renamed these to `Standard` and `VoiceEnabled` to make it a little more clear.